File: src/main/scala/ApiClient.scala
import com.typesafe.config.ConfigFactory
import dispatch._, Defaults._
import java.io.PrintWriter
import org.json4s.jackson.Serialization
import scala.concurrent.Future
import org.json4s._
import org.json4s.jackson.JsonMethods._


object ApiClient {

  val config = ConfigFactory.load()
  val authToken = config.getString("auth.token")
  val hostname = config.getString("host")
  implicit val formats = DefaultFormats


  private def buildRequest(path: String, data: Option[AnyRef]) = {
    val req = (host(hostname) / path <<? Map("auth" -> authToken)).POST
    if (data.isDefined) req << Serialization.write(data.get) else req
  }

  def makeRequest[TResp : Manifest](path: String, data: Option[AnyRef] = None): Future[Option[TResp]] = {
    val request = buildRequest(path, data)
    val future = Http(request OK as.String).option
    future.map { resp =>
      if (resp.isDefined)
        Some(parse(resp.get).extract[TResp])
      else
        None
    }
  }

  def Training(request: TrainRequest) = {
    makeRequest[TrainingProblem]("train", Some(request))
  }

  def Status() = {
    makeRequest[TrainingProblem]("status")
  }

  def Eval(request: EvalRequest) = {
    makeRequest[Status]("status", Some(request))
  }

  def Guess(request: Guess) = {
    makeRequest[GuessResponse]("guess", Some(request))
  }

  def Problems(writeToFile: Option[String] = None) = {
    val future = makeRequest[List[Problem]]("myproblems")
    future map {
      case Some(problems) => {
        writeToFile.map { x => Some(new PrintWriter(x)).map {
            writer => {
              val str = Serialization.write(problems)
              writer.write(str)
              writer.close()
            }
          }
        }
      }
      case None => ()
    }
    future
  }

}

File: fsharp/icfpc2013/icfpc2013/ApiClient.fs
module ApiClient
open ApiProtocol
open JsonSerializer
open System
open System.IO
open System.Net
open System.Text
open System.Threading

let baseUrl = "http://icfpc2013.cloudapp.net"
let authToken = "0201LxVGJ0tIY2MX5ce2dKYAU4b2NXYSXpOmNoQvvpsH1H"
let timeout = 1000

let rec request<'b, 'a> (path: string)(data: Option<'a>) =  
    let body = if data.IsSome then JsonSerializer.toJson data.Value else ""
    let bodyBytes = Encoding.Default.GetBytes(body)
    let builder = new UriBuilder(baseUrl)
    builder.Path <- path
    builder.Query <- "auth=" + authToken
    let req = HttpWebRequest.Create(builder.Uri) :?> HttpWebRequest
    req.Method <- "POST"
    req.ContentType <- "application/json"
    req.GetRequestStream().Write(bodyBytes, 0, bodyBytes.Length)
    req.GetRequestStream().Close()
    try
        let resp = req.GetResponse() :?> HttpWebResponse
        if resp.StatusCode = HttpStatusCode.OK then
            let respBody = (new StreamReader(resp.GetResponseStream())).ReadToEnd()
            JsonSerializer.fromJson<'b> respBody
        else
            failwith ("wrong status code: " + resp.StatusCode.ToString())
    with
        | :? WebException as e when e.Status = WebExceptionStatus.ProtocolError -> 
                Thread.Sleep(timeout)
                request path data

let train data: TrainingProblem = 
    request "train" (Some data)

let status: Status = 
    request "status" None

let eval data: EvalResponse = 
    request "eval" (Some data)

let guess data: GuessResponse =
    request "guess" (Some data)

let problems: array<Problem> = 
    request "myproblems" None

let cachedProblems =
    let json = File.ReadAllText("problems.json")
    JsonSerializer.fromJson<array<Problem>> json

open NUnit.Framework
open FsUnit
[<TestFixtureAttribute>]
type ``api client tests``() = 
    [<TestAttribute>]
    member x.TrainTest() = 
        let req = { TrainRequest.size = Some 3; operators = None }
        let result = train req
        result.size |> should equal req.size.Value
    [<TestAttribute>]
    member x.StatusTest() =
        let result = status
        result.contestScore |> should equal 0
    [<TestAttribute>]
    member x.ProblemsTest() =
        let problems = problems
        let solved = Array.filter (fun x -> x.solved.IsSome) problems
        let json = JsonSerializer.toJson problems
        File.WriteAllText("problems.json", json)
        problems.Length |> should equal 1820
    [<TestAttribute>]
    member x.CachedProblems() =
        let problems = cachedProblems
        problems.Length |> should equal 1820
                 

